### DISCUSSION TITLE
Questions about "fresh", db relationships, non-db models, caching, and .env

### DISCUSSION THREAD

**Discussion by quadrastreet** [leafsphp/leaf#61] [Q&A]
I'm brand new to Leaf, so I have a few questions (I'll be using leaf-api):

1) fresh() -- Given your example in the docs:

```
$flight = Flight::where('number', 'FR 900')->first();
$freshFlight = $flight->fresh();
```

If I didn't use fresh(), and simply re-ran the first line, would it not re-retrieve from the database?

2) Relationships -- Is there support for the idea of parent / child records using models? If so, how do I set it up?

3) Non-DB models -- In my API I'll need to be reading/writing using a database, as well as reading data from a 3rd-party REST API; I know I can put code in my controller to access data via 3rd-party REST API, but I was wondering if you had a strategy to use Leaf models where the data comes from that API rather than a database.

4) Caching -- Is there support for caching data rather than re-reading from the database with each request? If not, do you have any strategies that might be helpful to me?

5) .env -- I've read through https://leafphp.netlify.app/#/leaf-api/v/2.0/config/env but still have some questions:
-- APP_ENV -- How is this used?
-- APP_KEY -- How is this used?
-- APP_URL -- How is this used?

**Comment by quadrastreet** [comment]
Note: I found the answer to my 2nd question (model relationships) by reading the Eloquent Model documentation... once I saw the note that Leaf Model extends Eloquent Model.

**Reply by quadrastreet** [reply]
Note: after reading some more Eloquent and Laravel forums, I'm pretty sure I've figured out number 3 (re: non-db models): I'll need to write up custom code at the controller layer. However, if you have a strategy to recommend, that would be great.

**Reply by mychidarko** [reply]
Yeah, as mentioned in the [leaf docs](https://leafphp.netlify.app/#/leaf/v/2.5.0/core/model), leaf models are extended from eloquent, so using `Leaf\Model` or `Leaf\Database` is pretty much the same as using the using eloquent. You can follow the links on the docs page above to find out more if you're not already used to eloquent.

That would take care of questions 1, 2 and 3

**Reply by mychidarko** [reply]
For question 4, leaf does not come with any database caching out of the box, however, you can always implement this manually or check out an external library like [phpfastcache](https://github.com/PHPSocialNetwork/phpfastcache)

**Reply by mychidarko** [reply]
`APP_ENV` is used internally by leaf as an alternative to the [`mode` config](https://leafphp.netlify.app/#/leaf/v/2.5.0/config/settings?id=mode), however, as at this version `APP_KEY` and `APP_URL` are not used in your app. You might want to set it if you will need a reference to your app URL, say on a production server

**Reply by quadrastreet** [reply]
Thanks for all your thoughtful answers. I've been working on some caching techniques for particular sets of data, and I'll look a bit further into phpfastcache.

**Reply by mychidarko** [reply]
What was I thinking? 😂

Leaf actually does have some http caching out of the box https://leafphp.netlify.app/#/leaf/v/2.5.0/http/headers?id=http-caching-new

**Reply by quadrastreet** [reply]
Can you explain how to implement this is in my route?

Where I normally have something like this:

`$app->get("/info", "InformationController@info");`

I'm not sure how I use

```
$app->get("/info", function () use ($app) {
  Headers::etag("unique-tag");
  ...
});
```

**Reply by mychidarko** [reply]
In the `info` method in your controller, you can call `Leaf\Http\Headers::etag`, it's pretty much the same

**Reply by quadrastreet** [reply]
OK -- that seems to work.

**Reply by mychidarko** [reply]
Awesome
